### ISSUE TITLE
LinuxProcessInfoParser.parse not handling spaces in proc name

### ISSUE THREAD

**Issue by ebwweb** [jezhumble/javasysmon#10]
The parse method throws NumberFormatExceptions if the process name in the /proc/<pid>/stat file has spaces (for example the "MpxTestDaemon  " (with 2 spaces at the end) process on EMC/StorNext metadata controllers).

Recommend changing the parse method to:

``` java

    public ProcessInfo parse() {
        int openParen = stat.indexOf("(");
        int closeParen = stat.lastIndexOf(")");  

        //Start splitting after close of proc name
        String[] statElements = stat.substring(closeParen + 2).split(" "); 
        String pidStr = stat.substring(0,openParen - 1);

        return new ProcessInfo(Integer.parseInt(pidStr),
                Integer.parseInt(statElements[1]),
                trim(cmdline), getFirstMatch(STATUS_NAME_MATCHER, status),
                (String) uids.get(getFirstMatch(STATUS_UID_MATCHER, status)),
                Long.parseLong(statElements[11]) * (1000 / userHz),
                Long.parseLong(statElements[12]) * (1000 / userHz),
                Long.parseLong(getFirstMatch(STATUS_VM_RSS_MATCHER, status)) * 1024,
                Long.parseLong(getFirstMatch(STATUS_VM_SIZE_MATCHER, status)) * 1024);
    }
```

**Comment by jezhumble** [comment]
Thanks for the bug report.

The quickest way to get this in would be to fork, write a test, write
updated code that makes the test pass, and request a merge.

Jez.

On 22 August 2011 18:19, ebwweb <
reply@reply.github.com>wrote:

> The parse method throws NumberFormatExceptions if the process name in the
> /proc/<pid>/stat file has spaces (for example the "MpxTestDaemon  " (with 2
> spaces at the end) process on EMC/StorNext metadata controllers).
> 
> Recommend changing the parse method to:
> 
> ``` java
> 
>    public ProcessInfo parse() {
>        int openParen = stat.indexOf("(");
>        int closeParen = stat.lastIndexOf(")");
> 
>        //Start splitting after close of proc name
>        String[] statElements = stat.substring(closeParen + 2).split(" ");
>        String pidStr = stat.substring(0,openParen - 1);
> 
>        return new ProcessInfo(Integer.parseInt(pidStr),
>                Integer.parseInt(statElements[1]),
>                trim(cmdline), getFirstMatch(STATUS_NAME_MATCHER, status),
>                (String) uids.get(getFirstMatch(STATUS_UID_MATCHER,
> status)),
>                Long.parseLong(statElements[11]) * (1000 / userHz),
>                Long.parseLong(statElements[12]) * (1000 / userHz),
>                Long.parseLong(getFirstMatch(STATUS_VM_RSS_MATCHER, status))
> * 1024,
>                Long.parseLong(getFirstMatch(STATUS_VM_SIZE_MATCHER,
> status)) * 1024);
>    }
> ```
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jezhumble/javasysmon/issues/10

## 

Jez Humble
Co-author, _Continuous Delivery http://continuousdelivery.com/_
http://continuousdelivery.com/
http://jezhumble.net/

**Comment by jsankey** [comment]
Having recently adopted javasysmon, we had a customer report this issue.  I checked out the Linux source for parsing stat entries and they do it essentially as ebwweb has outlined above - by searching from both ends of the string to find the parens around the process name.  This is required as the name can include spaces, parens, numbers, pretty much anything.  I've adapted ebwweb's solution into a larger patch which includes tests and handling of various potentially unexpected conditions.  The aim is for things to be as lenient as possible (i.e. only require the fields we are trying to use) and to soldier on in the face of problem entries (i.e. skip difficult processes and continue gathering the rest).
